styx-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java

```diff
@@ -377,14 +398,18 @@ public void updateNextNaturalTrigger(WorkflowId workflowId, TriggerInstantSpec t
   }
 
   /**
-   * Eventually consistently list all active states and strongly consistently fetch their values.
-   *
-   * <p>This method will return a map of active states that might be missing some recently created
-   * states, but the values of all the states returned should be fresh.
+   * Strongly consistently read all active states
    */
   Map<WorkflowInstance, RunState> readActiveStates() throws IOException {
-    // Eventually consistently read active state keys
-    final List<Key> keys = readActiveInstanceKeys();
+    // Strongly read active state keys from index shards
+    final List<Key> keys = activeWorkflowInstanceIndexShardKeys(datastore.newKeyFactory()).stream()
+        .map(key -> forkJoinPool.submit(() ->
+            datastore.run(Query.newEntityQueryBuilder().setFilter(PropertyFilter.hasAncestor(key)).build())))
```

Comment:
Does it make sense to put kind as filter as well? Or filtering by ancestor does it implicitly?

Reply:
👍